src/rvsdg/schema.egg

```diff
@@ -48,6 +49,97 @@
+(rewrite (SubstExpr (eq ty a b) x v) (eq ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (lt ty a b) x v) (lt ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (gt ty a b) x v) (gt ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (le ty a b) x v) (le ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (ge ty a b) x v) (ge ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (not ty a b) x v) (not ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (and ty a b) x v) (and ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (or ty a b) x v) (or ty (SubstOperand a x v) (SubstOperand b x v)))
+
+(rewrite (SubstOperand (Arg x) x v) v)
+(rule ((= f (SubstOperand (Arg y) x v)) (!= y x))
+      ((union f (Arg y))))
+(rewrite (SubstOperand (Node b) x v) (Node (SubstBody b x v)))
+(rewrite (SubstOperand (Project i b) x v) (Project i (SubstBody b x v)))
+
+(rewrite (SubstBody (PureOp e) x v) (PureOp (SubstExpr e x v)))
+;; Subst doesn't cross regions - so we subst into the inputs but not outputs
+;; Node that a Gamma node's idx is on the outside, so it gets affected, but not
+;; a Theta node's predicate
+(rewrite (SubstBody (Gamma pred inputs outputs) x v) (Gamma (SubstOperand pred x v) (SubstVecOperand inputs x v) outputs))
+(rewrite (SubstBody (Theta pred inputs outputs) x v) (Theta pred (SubstVecOperand inputs x v) outputs))
+
+;; params: vec, var, op, next index to subst
+;; rtjoa: TODO: implement by mapping internally so they're not O(n^2) time
+(function SubstVecOperandHelper (VecOperand i64 Operand i64) VecOperand)
+(rewrite (SubstVecOperand vec x v) (SubstVecOperandHelper vec x v 0))
+(rule
+  (
+    (= f (SubstVecOperandHelper (VO vec) x v i))
+    (< i (vec-length vec))
+  )
+  ( 
+    (union
+      (SubstVecOperandHelper (VO vec) x v i)
+      (SubstVecOperandHelper
+        (VO (vec-set vec i (SubstOperand (vec-get vec i) x v)))
+        x v (+ i 1)
+        ))
+  ))
+
+(rule
+  (
+    (= f (SubstVecOperandHelper (VO vec) x v i))
+    (= i (vec-length vec))
+  )
+  ( 
+    (union (SubstVecOperandHelper (VO vec) x v i) (VO vec))
+  ))
+
+
+;; params: vec, var, op, next index to subst
+;; rtjoa: TODO: implement by mapping internally so they're not O(n^2) time
+(function SubstVecVecOperandHelper (VecVecOperand i64 Operand i64) VecVecOperand)
+(rewrite (SubstVecVecOperand vec x v) (SubstVecVecOperandHelper vec x v 0))
+(rule
+  (
+    (= f (SubstVecVecOperandHelper (VVO vec) x v i))
+    (< i (vec-length vec))
+  )
+  ( 
+    (union
+      (SubstVecVecOperandHelper (VVO vec) x v i)
+      (SubstVecVecOperandHelper
+        (VVO (vec-set vec i (SubstVecOperand (vec-get vec i) x v)))
+        x v (+ i 1)
+        ))
+  ))
+
+(rule
+  (
+    (= f (SubstVecVecOperandHelper (VVO vec) x v i))
+    (= i (vec-length vec))
+  )
+  ( 
+    (union (SubstVecVecOperandHelper (VVO vec) x v i) (VVO vec))
+  ))
 
```

Comment:
Can you add a ruleset for the subst rules to make it more usable?
I'd also like the existing egglog encoding to run this ruleset to saturation in-between runs of the normal ruleset.

Reply:
Added all rules to the `subst` ruleset, added the existing rules to constant-fold +/* to the `cfold` ruleset, and made the schedule the following:
```rs
let schedule = format!("(run-schedule (repeat {} cfold subst))", self.num_iters);
```